### DISCUSSION TITLE
Drop Data Stores?

### DISCUSSION THREAD

**Discussion by soininen** [Spine-project/Spine-Toolbox#1624] [Discussion]
I just wanted to float this idea here to shake up some discussion. Please don't take it too seriously!

Data Stores are a bit different project items to the rest of them. Their main purpose is to provide database URLs to other project items but they do database updates in some cases a well. This speciality has been a long source of issues on how to incorporate Data Stores into workflows, including backward resources, the Transformer item that doesn't actually transform anything and the discussions about the special 'resource link'.

I think the concept of Data Store as project item is fundamentally wrong and I have a radical proposal on how we could alleviate the situation.

How about we remove Data Store as a project item and replace it with a project-wide list of available databases?

I'm not talking about a new project item here, but a 'Database resources' list that would include all databases available to a Toolbox project. Project items that need a database URL would have source and/or destination database in their properties that points to a database in the list. Tools etc. would have list of all databases in their available command line arguments.

How would it work in practice? Consider the following DAG:

![drop-data-stores-before](https://user-images.githubusercontent.com/19147159/145542755-1f1e3c96-5b07-4283-a506-120496530063.png)

With this proposal, instead of the Input and Output data stores, the project would have a global list of the corresponding database URLs. Importer would be configured to write into the input database via its properties while Tool would get both input and output database URLs from its command line arguments. The DAG would then shrink to this:

![drop-data-stores-after](https://user-images.githubusercontent.com/19147159/145543017-f786f255-4f82-405a-9954-97428c9148f4.png)

As all items have access to all available databases, we wouldn't need Data Stores that actually point to the same database, backwards resources, or special 'resource links'. Data Transformer would become another database in the list, configured with the desired filters. We would need one new project item, though, a 'Database Writer', i.e. the resurrected Combiner.

Anyway, I'm not sure if we actually want to do some as radical as proposed here. I think it is still worthwhile to have this proposal out here in the open. Maybe it will spark some even nicer ideas?

**Comment by DillonJ** [comment]
You know what - I quite like this idea. I have had the same feelings about datastores...

I think your idea makes a lot of sense... one thing it's missing is the visual representation of what data is going where. My idea of resource links was to try and just capture this element, visualising the data flows as disctinct from the execution flows.

What about a hydrid solution. As you say, except that this project-wide list of datastores appears visually as, perhaps, smaller datastore icons in a fixed position to the side in the design area. You then get some sort of visual arrow to all the tools that use each "datastore" which would be the equivalent of the resource links... just to provide a visualisation of what data goes where.

But what about scenarios and tool filters? Perhaps they could operate on the resource links?

**Comment by manuelma** [comment]
I don't really like this idea I think. To me, the first workflow is much easier to follow than the second one. If you add arrows from fixed Data Stores on the side of the Desing view as @DillonJ is suggesting, you just end up with a less flexible interface.

I like to see the workflow with the Data Transformers and Data Stores (and Data Connections, which aren't considered in this proposal although they also provide data). Removing them from the workflow would be a loss in my humble opinion. I don't know what's so problematic about Data Stores at the moment, except that sometimes we execute them -- but that can be addressed by introducing the Combiner/Merger as discussed in another issue.

**Comment by jkiviluo** [comment]
I find it difficult to swallow as well. There are more complicated workflows where it's very good to directly see how the data flows and have a direct way to choose the right DB to be opened in DB editor. For example in TradeRES we will have:

- a single MySQL DB that holds all the data (fed by possibly multiple importers)
- multiple exporters from that central DB to model specific SQlite DBs (that are used by local modellers)
- then the models produce model specific outputs into DBs
- from which data is collected again to a single MySQL output DB

It would get unwieldy to just see a list of DBs.

It can still be a good idea to have a list of project DBs. Selecting one of them should highlight the DBs that it points to. Right now we have a list of Data Store items in the project 'explorer view', but it could actually be a list of DBs. And not all of them need to have active items.

**Comment by manuelma** [comment]
I forgot to say, even if I'm not a very big fan of this idea, I am a huge fan of thinking outside of the box like this. I see the issue that motivates this proposal, that DS (and even DC) are a bit special regarding execution. Maybe we can refine this proposal into something we all like?
